### PR TITLE
Set go version to 1.17 addendum

### LINKED ISSUE #9270: Update Go client to use 1.17
**Description**

Our Go client is now built against 1.15, which is not supported anymore. We should start building against Go 1.17, which is the lowest support version.

### PR DESCRIPTION
**Pull request by pihme**
## Description

- Sets go to 1.17 in the places I missed before
- Reads go version from `go.mod` file
- Enables caching for Go artifacts

**Review hints**
The last commit - enabling caching - is debatable. It doesn't add much, because the slow parts of the build are in the Java and Docker parts, not the go steps.

## Related issues

closes #9270

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
Seems like the flaky test could be fixed by reusing the same trick as in https://github.com/camunda/zeebe/pull/9502 :thinking:

File: .ci/pipelines/release_zeebe_java11.groovy
Comment (npepinpe) [inline .ci/pipelines/release_zeebe_java11.groovy:44]: :x: Please leave to 1.15.15 here, since the Java 11 release pipeline is only for versions < 1.3 (and these will not be updated to run on 1.17 AFAIK).

File: .github/workflows/code-quality.yml
Comment (npepinpe) [inline .github/workflows/code-quality.yml:41]: :+1:

**Review by npepinpe — APPROVED** [review]
:+1: 

Nice catch with the QA Testbench workflow, I completely forgot about that. I'm approving as there's no need for another review, but please have a look at the comments before merging

**Comment by pihme** [comment]
bors merge

**Pull request merged**
